docs/proposals/env-vars.md

```diff
@@ -0,0 +1,516 @@
+
+- making sure no entries in `DisallowedEnvironmentVariableKeys` exist in any `BuildStep` on the `BuildStrategy` when it is created
+- making sure no entries in `DisallowedEnvironmentVariableKeys` exist in the new environment variable fields proscribed by 
+this EP for the `Build` and `BuildRun` types.
+
+An alternative on the admission webhook approach, the current `pkg/validate/validate.go` code has some `Build`
+validations.  The interface proscribed there, along with the existing Strategy validation, could be augmented to ensure the 
+conditions noted above.  Expansion to consider `BuildRun` in that validation flow would be required.
+
+An open question at the top of this EP has been added for reaching a consensus around which validation path(s) to employ.
+
+Updates to `BuildSpec` and `BuildRunSpec` to allow for an array of `EnvironmentVariable` is then straight forward.
+
+With this combination of `BuildStrategySpec` (both a new `EnvironmentVariable` array and new string array called `DisallowedEnvironmentVariables` ),
+`BuildStep` (and the existing inlining of `Container`), and `EnvironmentVariable` arrays added to `BuildSpec` and 
+`BuildRunSpec`, we get an order of precedence that allows both "cluster admin centric" control over environment variables, but 
+opt in flexibility for "developer centric" overrides, on an `EnvironmentVariable` key by key basis.
+
+The order of precedence is:
+
+- `DisallowedEnvironmentVariableKeys` in the `BuildStrategySpec` is considered first.  Any use of those keys in 
+  `BuildStep`, `BuildStrategySpec.[]EnvironmentVariable`, `BuildSpec.[]EnvironmentVariable`, `BuildRunSpec.[]EnvironmentVariable` results in a validation error.
+- `BuildStep` from the `BuildStrategySpec` is next.  The key/value pair from these override anything from
+  `BuildStrategySpec.[]EnvironmentVariable`, `BuildSpec.[]EnvironmentVariable`, `BuildRunSpec.[]EnvironmentVariable`
+- `BuildRunSpec.[]EnvironmentVariable` is next
+- followed by `BuildSpec.[]EnvironmentVariable`
+- and lastly, `BuildStrategySpec.[]EnvironmentVariable`
+
+Or conversely, the environment variables are applied in this fashion
+- first from `BuildStrategySpec.[]EnvironmentVariable` assuming any names from `BuildStrategySpec.([]string)DisallowedEnvironmentVariable` are not used
+- then from `BuildSpec.[]EnvironmentVariable`, assuming any names from `BuildStrategySpec.([]string)DisallowedEnvironmentVariable` are not used,
+  where values are replaced from any entries previously seeded
+- then from `BuildRunSpec.[]EnvironmentVariable`, assuming any names from `BuildStrategySpec.([]string)DisallowedEnvironmentVariable` are not used,
+  where values are replaced from any entries previously seeded
+- then from `BuildStrategySpec.[]BuildStep`, assuming any names from `BuildStrategySpec.([]string)DisallowedEnvironmentVariable` are not used,
+  where values are replaced from any entries previously seeded  
```

Comment:
I was thinking of another approach:

* Everything specified in `BuildStrategySpec.[]BuildStep` may be overriden in Build/BuildRun unless specified otherwise in `.spec.[]DisallowEnvironmentVariableOverride`.

*  `.spec.[]DisallowEnvironmentVariableOverride` would contain the list of environment variables that a user is not allowed to be set irrespective of whether there was a default specified in `BuildStrategySpec.[]BuildStep`.
* `.spec.[]EnvironmentVariable` is not needed anymore because defaults could be specified in the `BuildStep` itself. 
 
( did you intend `BuildRunSpec.[]EnvironmentVariable` to be an allow-list of what env vars can be specified in build/buildRun ?)

Reply:
yep good question @sbose78 .... need some more  work here
- `BuildRunSpec.[]EnvironmentVariable` in intended to be defaults essentially
- but that said, admins might want to start with an "allow no envs, then list a small set of envs that can be used" vs. "allow all evns, then lis a small set of envs that cannot be used"

I'll add a `.spec.[]AllowedEnvironmentVariables` field, along with some thought / explanation on how `Allow...` and `Disallow...` work together.

Spoiler: `Allow...` and "allow no envs, then list a small set of envs that can be used" will take precedence.

Reply:
Agreed @gabemontero , sounds good!

Reply:
update pushed @sbose78